README.md
# element-within-viewport

[![Build Status][ci-img]][ci]
[![BrowserStack Status][browserstack-img]][browserstack]

Check if element is within viewport.

Convenient wrapper around [viewprt](https://github.com/gpoitch/viewprt)
supporting debounced listeners.

## Install

```sh
npm install element-within-viewport --save
```

## Usage

```js
import elementWithinViewport from 'element-within-viewport';

elementWithinViewport(document.querySelector('#jackie'), {
	onEnter: (element) => {
		// Element in viewport!
	}
});
```

## API

### elementWithinViewport(element, options)

Check if element is within viewport and calls proper callback.

#### element

Type: `Element`

Element to check.

#### options

Type: `object`

| Property              | Type       | Default value          | Description                                                                                                                                                                                          |
| --------------------- | ---------- | ---------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `threshold`           | `number`   | `0`                    | Positive value in pixels which will signal plugin to check for element presence earlier in document.                                                                                                 |
| `scrollResizeHandler` | `Function` | `(handler) => handler` | Window scroll and resize event handler. Useful if you want to use [throttle or debounce methods](#throttle-debounce-scroll-resize) on those events. Should return new handler (original or wrapped). |
| `onEnter`             | `Function` | `() => {}`             | Callback to execute if element is within viewport.                                                                                                                                                   |
| `onExit`              | `Function` | `() => {}`             | Callback to execute if element exits viewport.                                                                                                                                                       |
| `once`                | `boolean`  | `false`                | If true, will call `onEnter` only once.                                                                                                                                                              |

### instance.destroy()

Destroy instance.

## FAQ

### Throttling and debouncing scroll and resize event<a name="throttle-debounce-scroll-resize"></a>

If you want to throttle or debounce scroll and resize events, modify handler
with `scrollResizeHandler` property.

```js
import { debounce } from 'throttle-debounce';

elementWithinViewport(document.querySelector('#jackie'), {
	scrollResizeHandler: (handler) => debounce(300, handler)
});
```

## Browser support

Tested in Edge 15, Chrome 88 and Firefox 86, and should work in all modern
browsers.

## Test

For automated tests, run `npm run test:automated` (append `:watch` for watcher
support).

## License

MIT © [Ivan Nikolić](http://ivannikolic.com)

<!-- prettier-ignore-start -->

[ci]: https://travis-ci.com/niksy/element-within-viewport
[ci-img]: https://travis-ci.com/niksy/element-within-viewport.svg?branch=master
[browserstack]: https://www.browserstack.com/
[browserstack-img]: https://www.browserstack.com/automate/badge.svg?badge_key=bWM5U1R1ZU9DWXdacnFFV2prdVBHN05sZWlSQ0pwUzZoZkdtbXdMRUtscz0tLU9YOEJ0Z0xKZmg5aTJvd1F0TDQ4dFE9PQ==--b50bf3dc08e7e2401acdcd270693865d61221cf9

<!-- prettier-ignore-end -->


CHANGELOG.md
# Changelog

## [Unreleased][]

## [3.0.1][] - 2021-05-04

### Fixed

-   Properly resolve ES module imports

## [3.0.0][] - 2021-05-04

### Changed

-   Update viewprt

### Removed

-   **Breaking**: Drop IE support, supported browser is Edge 15+
-   Fallback support

## [2.0.0][] - 2020-02-20

### Changed

-   Abstract scroll and resize event handling, so events could be wrapped with
    more functionality (e.g. throttle)

### Added

-   `scrollResizeHandler` property

### Removed

-   `debounce` property

## [1.3.0][] - 2019-07-22

### Changed

-   Update viewprt

## [1.2.1][] - 2019-05-17

### Changed

-   Upgrade package

## [1.2.0][] - 2019-05-17

### Changed

-   Add support for `onExit` method
-   Upgrade package

## [1.1.1][] - 2019-01-28

-   Make package environment agnostic by using `global` instead of `window`

## [1.0.0][] - 2019-01-09

### Added

-   Initial stable implementation

<!-- prettier-ignore-start -->

[1.0.0]: https://github.com/niksy/element-within-viewport/tree/v1.0.0
[1.1.1]: https://github.com/niksy/element-within-viewport/compare/v1.0.0...v1.1.1
[1.2.0]: https://github.com/niksy/element-within-viewport/tree/v1.2.0
[1.2.1]: https://github.com/niksy/element-within-viewport/compare/v1.2.0...v1.2.1
[1.3.0]: https://github.com/niksy/element-within-viewport/tree/v1.3.0
[2.0.0]: https://github.com/niksy/element-within-viewport/tree/v2.0.0
[3.0.0]: https://github.com/niksy/element-within-viewport/tree/v3.0.0

<!-- prettier-ignore-end -->

[unreleased]:
	https://github.com/niksy/element-within-viewport/compare/v3.0.1...HEAD
[3.0.1]: https://github.com/niksy/element-within-viewport/tree/v3.0.1


karma.conf.js
'use strict';

const path = require('path');
const fs = require('fs');
const { default: resolve } = require('@rollup/plugin-node-resolve');
const commonjs = require('@rollup/plugin-commonjs');
const nodeBuiltins = require('rollup-plugin-node-builtins');
const globals = require('rollup-plugin-node-globals');
const { default: babel } = require('@rollup/plugin-babel');
const istanbul = require('rollup-plugin-istanbul');
const rollupConfig = require('./rollup.config');

let config;

const isCI =
	typeof process.env.CI !== 'undefined' && process.env.CI !== 'false';
const isPR =
	typeof process.env.TRAVIS_PULL_REQUEST !== 'undefined' &&
	process.env.TRAVIS_PULL_REQUEST !== 'false';
const local = !isCI || (isCI && isPR);

const port = 0;

if (local) {
	config = {
		browsers: ['Chrome']
	};
} else {
	config = {
		hostname: 'bs-local.com',
		browserStack: {
			username: process.env.BROWSER_STACK_USERNAME,
			accessKey: process.env.BROWSER_STACK_ACCESS_KEY,
			startTunnel: true,
			project: 'element-within-viewport',
			name: 'Automated (Karma)',
			build: 'Automated (Karma)'
		},
		customLaunchers: {
			'BS-Chrome': {
				base: 'BrowserStack',
				browser: 'Chrome',
				'browser_version': '88',
				os: 'Windows',
				'os_version': '7',
				project: 'element-within-viewport',
				build: 'Automated (Karma)',
				name: 'Chrome'
			},
			'BS-Firefox': {
				base: 'BrowserStack',
				browser: 'Firefox',
				'browser_version': '86',
				os: 'Windows',
				'os_version': '7',
				project: 'element-within-viewport',
				build: 'Automated (Karma)',
				name: 'Firefox'
			},
			'BS-Edge15': {
				base: 'BrowserStack',
				browser: 'Edge',
				'browser_version': '15',
				os: 'Windows',
				'os_version': '10',
				project: 'element-within-viewport',
				build: 'Automated (Karma)',
				name: 'Edge15'
			}
		},
		browsers: ['BS-Chrome', 'BS-Firefox', 'BS-Edge15']
	};
}

module.exports = function (baseConfig) {
	baseConfig.set({
		basePath: '',
		frameworks: ['mocha', 'viewport', 'fixture'],
		files: [
			'test/**/*.css',
			'test/**/*.html',
			{ pattern: 'test/**/*.js', watched: false }
		],
		exclude: [],
		preprocessors: {
			'test/**/*.html': ['html2js'],
			'test/**/*.js': ['rollup', 'sourcemap']
		},
		reporters: ['mocha', 'coverage-istanbul'],
		port: port,
		colors: true,
		logLevel: baseConfig.LOG_INFO,
		autoWatch: false,
		client: {
			mocha: {
				timeout: 20000
			},
			captureConsole: true
		},
		browserConsoleLogOptions: {
			level: 'log',
			format: '%b %T: %m',
			terminal: true
		},
		browserNoActivityTimeout: 60000,
		rollupPreprocessor: {
			plugins: [
				nodeBuiltins(),
				babel({
					exclude: 'node_modules/**',
					babelHelpers: 'runtime'
				}),
				resolve({
					preferBuiltins: true
				}),
				commonjs(),
				babel({
					include: 'node_modules/{has-flag,supports-color}/**',
					babelHelpers: 'runtime',
					babelrc: false,
					configFile: path.resolve(__dirname, '.babelrc')
				}),
				globals(),
				...rollupConfig.plugins.filter(
					({ name }) => !['babel'].includes(name)
				),
				istanbul({
					exclude: ['test/**/*.js', 'node_modules/**/*']
				})
			],
			output: {
				format: 'iife',
				name: 'elementWithinViewport',
				sourcemap: baseConfig.autoWatch ? false : 'inline', // Source map support has weird behavior in watch mode
				intro: 'window.TYPED_ARRAY_SUPPORT = false;' // IE9
			}
		},
		coverageIstanbulReporter: {
			dir: path.join(__dirname, 'coverage/%browser%'),
			fixWebpackSourcePaths: true,
			reports: ['html', 'text'],
			thresholds: {
				global: JSON.parse(
					fs.readFileSync(path.join(__dirname, '.nycrc'), 'utf8')
				)
			}
		},
		singleRun: true,
		concurrency: 1,
		...config
	});
};
